File: dart_lang/lib/c_strategy/b_enhanced_solution/product.dart
import 'package:dart_lang/c_strategy/b_enhanced_solution/pricing_strategy/pricing_strategy_interface.dart';

class Product {
  final String _name;
  final double _price;
  final PricingStrategy _pricingStrategy;

  const Product(
      {required String name,
      required double price,
      required PricingStrategy pricingStrategy})
      : _name = name,
        _price = price,
        _pricingStrategy = pricingStrategy;

  String get name => _name;
  double get price => _price;

  double calculatePrice() {
    return _pricingStrategy.calculatePrice(_price);
  }
}

File: PHP/Patterns/src/Behavioral/Strategy/Models/Product.php
<?php

namespace DesignPatterns\Behavioral\Strategy\Models;

use DesignPatterns\Behavioral\Strategy\Interfaces\PricingStrategy;

class Product
{
    private PricingStrategy $pricingStrategy;
    private float $price;
    private string $name;

    public function __construct(float $price, string $name, PricingStrategy $pricingStrategy)
    {
        $this->pricingStrategy = $pricingStrategy;
        $this->price = $price;
        $this->name = $name;
    }

    public function getPrice(): float
    {
        return $this->pricingStrategy->calculatePrice($this->price);
    }

    public function getName(): string
    {
        return $this->name;
    }
}
